src/main/java/com/wanderersoftherift/wotr/world/level/levelgen/processor/AttachmentProcessor.java

```diff
@@ -0,0 +1,145 @@
+                    Codec.BOOL.optionalFieldOf("requires_up", false).forGetter(AttachmentProcessor::isRequiresUp),
+                    Codec.BOOL.optionalFieldOf("requires_down", false).forGetter(AttachmentProcessor::isRequiresDown),
+                    Codec.FLOAT.fieldOf("rarity").forGetter(AttachmentProcessor::getRarity),
+                    RANDOM_TYPE_CODEC.optionalFieldOf("random_type", StructureRandomType.BLOCK).forGetter(AttachmentProcessor::getStructureRandomType),
+                    Codec.LONG.optionalFieldOf("seed", 7645816L).forGetter(AttachmentProcessor::getSeed)
+            ).apply(builder, AttachmentProcessor::new));
+
+    private final BlockState blockState;
+    private final int requiresSides;
+    private final boolean requiresUp;
+    private final boolean requiresDown;
+    private final float rarity;
+    private final StructureRandomType structureRandomType;
+    private final long seed;
+
+    public AttachmentProcessor(BlockState blockState, int requiresSides, boolean requiresUp, boolean requiresDown, float rarity, StructureRandomType structureRandomType, long seed) {
+        this.blockState = blockState;
+        this.requiresSides = requiresSides;
+        this.requiresUp = requiresUp;
+        this.requiresDown = requiresDown;
+        this.rarity = rarity;
+        this.structureRandomType = structureRandomType;
+        this.seed = seed;
+    }
+
+    @Override
+    public List<StructureTemplate.StructureBlockInfo> finalizeProcessing(ServerLevelAccessor serverLevel, BlockPos offset, BlockPos pos, List<StructureTemplate.StructureBlockInfo> originalBlockInfos, List<StructureTemplate.StructureBlockInfo> processedBlockInfos, StructurePlaceSettings settings) {
+        //Map<BlockPos, StructureTemplate.StructureBlockInfo> mapByPos = mapByPos(processedBlockInfos);
+        List<StructureTemplate.StructureBlockInfo> newBlockInfos = new ArrayList<>();
+        for (StructureTemplate.StructureBlockInfo blockInfo : processedBlockInfos) {
+            StructureTemplate.StructureBlockInfo newBlockInfo = processFinal(serverLevel, offset, pos, blockInfo, blockInfo, settings, processedBlockInfos);
+            newBlockInfos.add(newBlockInfo);
+        }
+        return newBlockInfos;
+    }
+
+
+    public StructureTemplate.StructureBlockInfo processFinal(LevelReader world, BlockPos piecePos, BlockPos structurePos, StructureTemplate.StructureBlockInfo rawBlockInfo, StructureTemplate.StructureBlockInfo blockInfo, StructurePlaceSettings settings, List<StructureTemplate.StructureBlockInfo> processedBlockInfos) {
+        RandomSource random = ProcessorUtil.getRandom(structureRandomType, blockInfo.pos(), piecePos, structurePos, world, seed);
+        BlockState blockstate = blockInfo.state();
+        BlockPos blockpos = blockInfo.pos();
+        if(blockstate.isAir() && random.nextFloat() <= rarity){
+            boolean hasSides = isHasSides(blockpos, processedBlockInfos);
```

Comment:
Niggly, but these booleans might be better named `validSides`, `validUp`, `validDown` since they will be true if not required too.